pkg/controller/provider/xavier/base/client.go

```diff
@@ -0,0 +1,335 @@
+	Cluster         = "cluster"
+	VmName          = "vmName"
+	DiskSpace       = "diskSpace"
+	Memory          = "memory"
+	CpuCores        = "cpuCores"
+	GuestOs         = "guestOSFullName"
+	OsProduct       = "osProductName"
+	Product         = "product"
+	Version         = "version"
+	HostName        = "host_name"
+	BalloonedMemory = "balloonedMemory"
+	HasMemoryHotAdd = "hasMemoryHotAdd"
+	HasCpuHotAdd    = "hasCpuHotAdd"
+	HasCpuHotRemove = "hasCpuHotRemove"
+	HasCpuAffinity  = "cpuAffinityNotNull"
+	HasDrsEnabled   = "hasVmDrsConfig"
+	HasHaEnabled    = "hasVmHaConfig"
+	HasPassthrough  = "hasPassthroughDevice"
+	HasUsb          = "hasUSBcontrollers"
+	HasRdmDisk      = "hasRdmDisk"
+	ScanDate        = "scanRunDate"
+)
+
+//
+// Application settings.
+var Settings = &settings.Settings
+
+type List = []interface{}
+
+type Object = map[string]interface{}
+
+//
+// Model adapter.
+type ModelAdapter interface {
+	// Update concerns.
+	UpdateConcerns(model libmodel.Model, concerns []string)
+	// Update the xavier body.
+	UpdateBody(object Object, model libmodel.Model) error
+}
+
+//
+// Factory.
+func New(db libmodel.DB, adapter ModelAdapter) *Client {
+	return &Client{
+		DB:       db,
+		URL:      Settings.Inventory.Xavier.URL,
+		User:     Settings.Inventory.Xavier.User,
+		Password: Settings.Inventory.Xavier.Password,
+		Adapter:  adapter,
+	}
+}
+
+//
+// Xavier client.
+type Client struct {
+	// Database.
+	libmodel.DB
+	// Service URL.
+	URL string
+	// User.
+	User string
+	// Password
+	Password string
+	// Builder
+	Adapter ModelAdapter
+}
+
+//
+// Analyze the specified VM.
+func (r *Client) Analyze(model libmodel.Model) (err error) {
+	if r.URL == "" {
+		Log.Info("Xavier: URL not configured.")
+		return
+	}
+	if mX, cast := model.(interface{ Analyzed() bool }); cast {
+		if mX.Analyzed() {
+			return
+		}
+	}
+	reply := Object{}
+	body := r.body(model)
+	status, err := r.post(&body, &reply)
+	if status != http.StatusOK {
+		err = liberr.New(http.StatusText(status))
+		return
+	}
+	report, err := r.find(
+		reply,
+		"result",
+		"execution-results",
+		"results",
+		"2",
+		"value",
+		"org.drools.core.runtime.rule.impl.FlatQueryResults",
+		"idResultMaps",
+		"element",
+		"0",
+		"element",
+		"0",
+		"value",
+		"org.jboss.xavier.analytics.pojo.output.workload.inventory.WorkloadInventoryReportModel")
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	providers, err := r.find(report, "recommendedTargetsIMS")
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	flags, err := r.find(report, "flagsIMS")
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	concerns := []string{}
+	if list, cast := flags.(List); cast {
+		for _, flag := range list {
+			concerns = append(concerns, flag.(string))
+		}
+	}
+	if len(concerns) == 0 {
+		recommended := false
+		if list, cast := providers.(List); cast {
+			for _, provider := range list {
+				if provider == "OCP" {
+					recommended = true
+					break
+				}
+			}
+		}
```

Comment:
Rules are triggered by _known_ field values.  So if, for example, an unknown guest OS is passed, some of the rules don't fire resulting in `OCP` not being in the recommendTargetsIMS but flagsIMS = nil.  This added as a catch-all.